ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs

```diff
@@ -0,0 +1,695 @@
+  -> LMDBBackingStoreInitialiser l m
+newLMDBBackingStoreInitialiser dbTracer limits = HD.BackingStoreInitialiser $ \sfs initFrom -> do
+  Trace.traceWith dbTracer TDBOpening
+
+  dbOpenHandles <- IOLike.newTVarIO Map.empty
+  dbClosed      <- IOLike.newTVarIO False
+
+  let
+    path = FS.mkFsPath ["tables"]
+
+    (gdd, copyDbAction :: m (), initAction :: Db m l -> m ()) = case initFrom of
+      HD.InitFromCopy (HD.BackingStorePath fp)
+        -- If fp == path then this is the LINoInitialise case
+        | fp /= path -> (GDDMustNotExist, initFromLMDBs dbTracer limits sfs fp path, \_ -> pure ())
+      HD.InitFromValues slot vals -> (GDDMustNotExist, pure (), initFromVals slot vals)
+      _ -> (GDDMustExist, pure (), \_ -> pure ())
+
+  -- get the filepath for this db creates the directory if appropriate
+  dbFilePath <- guardDbDirWithRetry gdd sfs path
+
+  -- copy from another lmdb path if appropriate
+  copyDbAction
+
+  -- open this database
+  dbEnv <- liftIO $ LMDB.openEnvironment dbFilePath (unLMDBLimits limits)
+
+  -- Create the settings table.
+  dbSettings <- liftIO $ LMDB.readWriteTransaction dbEnv $ LMDB.getDatabase (Just "_dbsettings")
+  liftIO $ LMDB.readWriteTransaction dbEnv $ writeDbSettings dbSettings
+
+  -- The LMDB.Database that holds the @`DbState`@ (i.e. sequence number)
+  -- This transaction must be read-write because on initialisation it creates the database
+  dbState <- liftIO $ LMDB.readWriteTransaction dbEnv $ LMDB.getDatabase (Just "_dbstate")
+
+  -- TODO: at some point Javier was able to get an LMDB which didn't have this
+  -- database. How is it possible? maybe some copy function is not copying this
+  -- one?
+
+  -- Here we get the LMDB.Databases for the tables of the ledger state
+  -- Must be read-write transaction because tables may need to be created
+  dbBackingTables <- liftIO $ LMDB.readWriteTransaction dbEnv $
+    traverseLedgerTables getDb namesLedgerTables
+
+  dbNextId <- IOLike.newTVarIO 0
+
+  let
+    db = Db{..}
+    bsClose :: m ()
+    bsClose = do
+      guardDbClosed dbClosed
+      Trace.traceWith dbTracer $ TDBClosing dbFilePath
+      openHandles <- IOLike.readTVarIO dbOpenHandles
+      for_ openHandles HD.bsvhClose
+      IOLike.atomically $ IOLike.modifyTVar dbClosed (const True)
+      liftIO $ LMDB.closeEnvironment dbEnv
+      Trace.traceWith dbTracer $ TDBClosed dbFilePath
+
+    bsCopy shfs (HD.BackingStorePath to0) = do
+      guardDbClosed dbClosed
+      to <- guardDbDir GDDMustNotExist shfs to0
+      lmdbCopy dbTracer dbEnv to
+
+    bsValueHandle =
+      guardDbClosed dbClosed >>
+      mkLMDBBackingStoreValueHandle db
+
+    bsWrite :: SlotNo -> LedgerTables l DiffMK -> m ()
+    bsWrite slot diffs = do
+      guardDbClosed dbClosed
+      oldSlot <- liftIO $ LMDB.readWriteTransaction dbEnv $ withDbStateRW dbState $ \s@DbState{dbsSeq} -> do
+        -- TODO This should be <. However the test harness does call bsWrite with the same slot
```

Comment:
@jorisdral What about this one?

Reply:
I think it is the same problem as https://github.com/input-output-hk/ouroboros-network/pull/4341/files#r1101447839. https://github.com/input-output-hk/ouroboros-network/pull/4342/files#diff-612ea931bbb4919adeba7cbf5f3de114eb48f9102be69b08d23fc44ef53c2afeR126 also has the same non-strict inequality.

I have updated  #4367 to also reference the strict inequalities in the `TVar` and LMDB backing store implementations.

Reply:
Closing this thread then